homeassistant/components/stream/__init__.py

```diff
@@ -0,0 +1,149 @@
+"""
+Provide functionality to stream video source.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/stream/
+"""
+import logging
+import threading
+
+import voluptuous as vol
+
+from homeassistant.auth.util import generate_secret
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import callback
+from homeassistant.exceptions import HomeAssistantError
+import homeassistant.helpers.config_validation as cv
+from homeassistant.loader import bind_hass
+
+from .const import (
+    DOMAIN, CONF_KEEPALIVE, ATTR_STREAMS, ATTR_ENDPOINTS)
+from .worker import stream_worker
+from .hls import async_setup_hls
+
+REQUIREMENTS = ['av==6.1.2']
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['http']
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_KEEPALIVE, default=False): cv.boolean,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+@bind_hass
+def request_stream(hass, stream_source, *, fmt='hls',
+                   options=None, preload=False):
+    """Set up stream with token."""
+    if DOMAIN not in hass.config.components:
+        raise HomeAssistantError("Stream component is not set up.")
+
+    if options is None:
+        options = {}
+
+    keepalive = hass.data[DOMAIN][CONF_KEEPALIVE]
+    try:
+        streams = hass.data[DOMAIN][ATTR_STREAMS]
+        stream = streams.get(stream_source)
+        if not stream:
+            stream = Stream(hass, stream_source,
+                            options=options, keepalive=keepalive)
+            streams[stream_source] = stream
+        if not preload and not stream.access_token:
+            stream.access_token = generate_secret()
+            stream.start()
+        return hass.data[DOMAIN][ATTR_ENDPOINTS][fmt].format(
+            hass.config.api.base_url, stream.access_token)
+    except Exception:
+        raise HomeAssistantError('Unable to get stream')
+
+
+@bind_hass
+def get_stream(hass, stream_source):
+    """Return available stream."""
+    streams = hass.data[DOMAIN][ATTR_STREAMS]
+    return streams.get(stream_source)
+
+
+async def async_setup(hass, config):
+    """Set up stream."""
+    conf = config[DOMAIN] if config.get(DOMAIN, {}) else {}
+
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][CONF_KEEPALIVE] = conf.get(CONF_KEEPALIVE)
+    hass.data[DOMAIN][ATTR_ENDPOINTS] = {}
+    hass.data[DOMAIN][ATTR_STREAMS] = {}
+
+    # Setup HLS
+    hls_endpoint = await async_setup_hls(hass)
+    hass.data[DOMAIN][ATTR_ENDPOINTS]['hls'] = hls_endpoint
+
+    @callback
+    def shutdown(event):
+        """Stop all stream workers."""
+        for stream in hass.data[DOMAIN][ATTR_STREAMS].values():
+            stream.keepalive = False
```

Comment:
Is this something that should be moved into `stream.stop()` ?

Reply:
This came from refactoring `stop` and `stop_stream` to `stop` and `_stop`.  `_stop` kills the thread, but `stop` only kills the thread if `keepalive` is false.  I can remove this and call `_stop` here on the shutdown, but would have to disable linting around calling a protected method.

Reply:
Maybe we rename then `stop` and `quit`  ?

Reply:
Actually it's fine like this too.

Reply:
Something like `def stop(force=False)` and `if not self.keepalive or force:` to call `_stop`?